File: src/main/java/com/pedigree/editor/SelectionModel.java
package com.pedigree.editor;

import java.lang.reflect.Method;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;

/**
 * Tracks selected entity IDs. Accepts either a String ID or an object exposing getId().
 */
public class SelectionModel {
    public interface BoundsProvider {
        /**
         * Returns a bounding rectangle (in layout coordinates) for a given node id.
         */
        Rect getBounds(String id);

        /**
         * Returns all selectable node ids.
         */
        Set<String> getAllIds();
    }

    private final LinkedHashSet<String> selectedIds = new LinkedHashSet<>();
    private BoundsProvider boundsProvider;

    public void setBoundsProvider(BoundsProvider provider) {
        this.boundsProvider = provider;
    }

    public void select(Object obj) {
        Objects.requireNonNull(obj, "obj");
        String id = resolveId(obj);
        selectedIds.clear();
        selectedIds.add(id);
    }

    public void addToSelection(Object obj) {
        Objects.requireNonNull(obj, "obj");
        selectedIds.add(resolveId(obj));
    }

    public void selectRange(Rect rect) {
        if (rect == null || boundsProvider == null) return;
        selectedIds.clear();
        for (String id : boundsProvider.getAllIds()) {
            Rect b = boundsProvider.getBounds(id);
            if (b != null && rect.intersects(b)) {
                selectedIds.add(id);
            }
        }
    }

    public void clear() {
        selectedIds.clear();
    }

    public boolean isSelected(String id) {
        return selectedIds.contains(id);
    }

    public Set<String> getSelectedIds() {
        return Collections.unmodifiableSet(selectedIds);
    }

    private static String resolveId(Object obj) {
        if (obj instanceof String s) return s;
        try {
            Method m = obj.getClass().getMethod("getId");
            Object val = m.invoke(obj);
            if (val instanceof String s) return s;
        } catch (Exception ignored) { }
        throw new IllegalArgumentException("Selection requires String id or object exposing getId()");
    }
}

File: src/main/kotlin/com/family/tree/editor/SelectionModel.kt
package com.family.tree.editor

import java.lang.reflect.Method
import java.util.*

/**
 * Tracks selected entity IDs. Accepts either a String ID or an object exposing getId().
 */
class SelectionModel {
    interface BoundsProvider {
        /**
         * Returns a bounding rectangle (in layout coordinates) for a given node id.
         */
        fun getBounds(id: String): Rect?

        /**
         * Returns all selectable node ids.
         */
        fun getAllIds(): Set<String>
    }

    private val selectedIds: LinkedHashSet<String> = LinkedHashSet()
    private var boundsProvider: BoundsProvider? = null

    fun setBoundsProvider(provider: BoundsProvider?) {
        this.boundsProvider = provider
    }

    fun select(obj: Any) {
        Objects.requireNonNull(obj, "obj")
        val id = resolveId(obj)
        selectedIds.clear()
        selectedIds.add(id)
    }

    fun addToSelection(obj: Any) {
        Objects.requireNonNull(obj, "obj")
        selectedIds.add(resolveId(obj))
    }

    fun selectRange(rect: Rect?) {
        if (rect == null) return
        val provider = boundsProvider ?: return
        selectedIds.clear()
        for (id in provider.getAllIds()) {
            val b = provider.getBounds(id)
            if (b != null && rect.intersects(b)) {
                selectedIds.add(id)
            }
        }
    }

    fun clear() {
        selectedIds.clear()
    }

    fun isSelected(id: String): Boolean = selectedIds.contains(id)

    fun getSelectedIds(): Set<String> = Collections.unmodifiableSet(selectedIds)

    private fun resolveId(obj: Any): String {
        if (obj is String) return obj
        try {
            val m: Method = obj.javaClass.getMethod("getId")
            val valObj = m.invoke(obj)
            if (valObj is String) return valObj
        } catch (_: Exception) {
        }
        throw IllegalArgumentException("Selection requires String id or object exposing getId()")
    }
}
